wcfsetup/install/files/lib/acp/form/LanguageAddForm.class.php

```diff
@@ -103,149 +73,105 @@ public function readParameters()
     /**
-     * Validates given source language.
+     * @inheritDoc
      */
-    protected function validateSource()
+    public function buildForm()
     {
-        if (empty($this->sourceLanguageID)) {
-            throw new UserInputException('sourceLanguageID');
-        }
-
-        // get language
-        $this->sourceLanguage = LanguageFactory::getInstance()->getLanguage($this->sourceLanguageID);
-        if (!$this->sourceLanguage->languageID) {
-            throw new UserInputException('sourceLanguageID');
-        }
+        parent::buildForm();
+
+        $this->form->getDataHandler()->addProcessor(new VoidFormDataProcessor('sourceLanguageID'));
+        $this->form->getDataHandler()->addProcessor(new CustomFormDataProcessor(
+            'lowercase',
+            static function (IFormDocument $document, array $parameters) {
+                $parameters['data']['languageCode'] = \mb_strtolower($parameters['data']['languageCode']);
+                $parameters['data']['countryCode'] = \mb_strtolower($parameters['data']['countryCode']);
+
+                return $parameters;
+            }
+        ));
     }
 
-    /**
-     * @inheritDoc
-     */
-    public function save()
+    public function saved()
     {
-        parent::save();
-
-        $this->language = LanguageEditor::create(\array_merge($this->additionalFields, [
-            'countryCode' => \mb_strtolower($this->countryCode),
-            'languageName' => $this->languageName,
-            'languageCode' => \mb_strtolower($this->languageCode),
-            'locale' => $this->locale,
-        ]));
-        $languageEditor = new LanguageEditor($this->sourceLanguage);
-        $languageEditor->copy($this->language);
-
-        // copy content
-        LanguageEditor::copyLanguageContent($this->sourceLanguage->languageID, $this->language->languageID);
-
-        // reset caches
+        // reset cache
         LanguageFactory::getInstance()->clearCache();
-        LanguageFactory::getInstance()->deleteLanguageCache();
 
-        $this->saved();
+        if ($this->formAction === IFormDocument::FORM_MODE_CREATE) {
+            /** @var Language $language */
+            $language = $this->objectAction->getReturnValues()['returnValues'];
 
-        // show success message
-        WCF::getTPL()->assign([
-            'success' => true,
-            'objectEditLink' => LinkHandler::getInstance()->getControllerLink(
-                LanguageEditForm::class,
-                ['id' => $this->language->languageID]
-            ),
-        ]);
-
-        // reset values
-        $this->countryCode = $this->languageCode = $this->languageName = $this->locale = '';
-        $this->sourceLanguageID = 0;
-    }
-
-    /**
-     * @inheritDoc
-     */
-    public function readData()
-    {
-        parent::readData();
+            // copy content
+            /** @var SingleSelectionFormField $sourceIDField */
+            $sourceIDField = $this->form->getNodeById('sourceLanguageID');
```

Comment:
```suggestion
            $sourceIDField = $this->form->getNodeById('sourceLanguageID');
            \assert($sourceIDField instanceof SingleSelectionFormField);
```

`assert()` achieves the same, but is also checked by PHP itself. The same for `$language`, please.